src/databricks/sql/client.py

```diff
@@ -297,6 +300,117 @@ def _check_not_closed(self):
+                raise Error("Local file operations are restricted to paths within the configured uploads_base_path")
+
+        # TODO: Experiment with DBR sending real headers.
+        # The specification says headers will be in JSON format but the current null value is actually an empty list []
+        handler_args = {
+            "presigned_url": row.presignedUrl,
+            "local_file": getattr(row, "localFile", None),
+            "headers": json.loads(row.headers or "{}"),
+        }
+
+        logger.debug(
+            f"Attempting staging operation indicated by server: {row.operation} - {getattr(row, 'localFile', '')}"
+        )
+
+        # TODO: Create a retry loop here to re-attempt if the request times out or fails
+        if row.operation == "GET":
+            return self._handle_staging_get(**handler_args)
+        elif row.operation == "PUT":
+            return self._handle_staging_put(**handler_args)
+        elif row.operation == "REMOVE":
+            # Local file isn't needed to remove a remote resource
+            handler_args.pop("local_file")
+            return self._handle_staging_remove(**handler_args)
+        else:
+            raise Error(
+                f"Operation {row.operation} is not supported. "
+                + "Supported operations are GET, PUT, and REMOVE"
+            )
+
+    def _handle_staging_put(
+        self, presigned_url: str, local_file: str, headers: dict = None
+    ):
+        """Make an HTTP PUT request
+
+        Raise an exception if request fails. Returns no data.
+        """
+
+        if local_file is None:
+            raise Error("Cannot perform PUT without specifying a local_file")
+
+        with open(local_file, "rb") as fh:
+            r = requests.put(url=presigned_url, data=fh, headers=headers)
+
+        # fmt: off
+        # Design borrowed from: https://stackoverflow.com/a/2342589/5093960
+            
+        OK = requests.codes.ok                  # 200
+        CREATED = requests.codes.created        # 201
+        ACCEPTED = requests.codes.accepted      # 202
+        NO_CONTENT = requests.codes.no_content  # 204
+
+        # fmt: on
+
+        if r.status_code not in [OK, CREATED, NO_CONTENT, ACCEPTED]:
+            raise Error(
+                f"Staging operation over HTTP was unsuccessful: {r.status_code}-{r.text}"
+            )
+
+        if r.status_code == ACCEPTED:
+            logger.debug(
+                f"Response code {ACCEPTED} from server indicates ingestion command was accepted "
+                + "but not yet applied on the server. It's possible this command may fail later."
+            )
+
+    def _handle_staging_get(
+        self, local_file: str, presigned_url: str, headers: dict = None
+    ):
+        """Make an HTTP GET request, create a local file with the received data
+
+        Raise an exception if request fails. Returns no data.
+        """
+
+        with open(local_file, "wb") as fp:
```

Comment:
nit: would it be better to open an outputStream after we get a successful response?